enhancements/cluster-logging/cluster-logging-move-init-to-es-operator.md

```diff
@@ -0,0 +1,72 @@
+---
+title: cluster-logging-log-forwarding
+authors:
+  - "@jcantrill"
+  - "@igor-karpukhin"
+  - "@ewolinetz"
+reviewers:
+  - "@bparees"
+  - "@jcantrill"
+  - "@igor-karpukhin"
+  - "@alanconway"
+  - "@ewolinetz"
+approvers:
+  - "@ewolinetz"
+  - "@jcantrill"
+creation-date: 2019-09-17
+last-updated: 2019-10-25
+status: implementable
+see-also:[]
+replaces:[]
+superseded-by:[]
+---
+
+# cluster-logging-es-init
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+The purpose of moving Elasticsearch initialization logic to the elasticsearch-operator is to reduce the number of times these init scripts executed.
+At the moment we keep these scripts in the ES image, and execute them everytime when ES pod starts. 
+
+
+### Goals
+The specific goals of this proposal are:
+
+* Provide an API that allows users to specify init procedures for the ES cluster
+* Allow users to choose what command has to be executed during the init process for FULL and ROLLING elasticsearch cluster restart
+
+## Proposal
+
+The following changes will define how the elasticsearch-operator executes init process for Elasticsearch cluster restarts
+
+
+* Adding additional section to the elasticsearch-operator called `actions`, which may look like this:
+
+    ```json
+    "actions": {
+      "image": "quay.io/openshift/origin-logging-elasticsearch6:latest",
+      "when": {
+          "postNew": "init.sh"
+          "postUpgrade": "init_upgrade.sh"
+      }
+    },
+    ```
+  The values for actions are commands to be executed from the image
+  * `postNew` - execute command after starting a cluster
+  * `postUpgrade` - execute command after a cluster upgrade
+  
+  If this section is not defined at all, nothing will be executed.
+  
+  If the `image` is provided by the user, `actions` section also has to be filled
+  with both actions and their commands
+   
+* Building an image out of these init scripts
+* Executing given scripts with the image that contains all the init logic against the ES API node
```

Comment:
this actually brings up a good point. should we also outline a structure for what the scripts should look like?
e.g. 
1. named variables that the operator will inject so that we can be sure to point the initialization logic to the correct cluster? 
    * the es host
    * cert paths
2. established return codes so the operator knows if it was successful
etc

Reply:
@ewolinetz good point. We may need to create an example script and tell the users that scripts should follow our defined structure. So that CLO will only execute that command and verify that it returned with code `0`.

Reply:
In the case where the logic is executed from outside the ES cluster I think you are right in that we may need to be able to provide arguments to the script for certs and destination.  Essentially we will need to do 'docker run' and pass i those things.  I think return codes are binary: 0 or not 0

Reply:
@ewolinetz and I had a conversation about that. We gonna have a couple of ENV variables for the init script. 
* `ES_HOST` - will point to ES master
* `CA_PATH` - will point to mounted CA secrets
* `ES_CERT` - will point to ES certificate
* `ES_KEY` - will point to ES key.

The `Job` containing init scripts will be executed against the ES master. Current init scripts will have to be adjusted.

Reply:
@igor-karpukhin i was thinking more like:
`ES_URL` `CA_PATH` `CLIENT_CERT_PATH` `CLIENT_KEY_PATH`

Reply:
@ewolinetz thanks. I added them to the proposal.